examples/water/main.rs

```diff
@@ -0,0 +1,797 @@
+            label: None,
+            size: texture_extent,
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Depth32Float,
+            usage: wgpu::TextureUsage::SAMPLED | wgpu::TextureUsage::COPY_DST | wgpu::TextureUsage::OUTPUT_ATTACHMENT,
+        });
+
+        self.draw_depth_buffer = draw_depth_buffer.create_default_view();
+        self.reflect_view = reflection_texture.create_default_view();
+
+        let (
+            normal_view,
+            flipped_view,
+            projection,
+        ) = Self::generate_matrices(sc_desc.width as f32 / sc_desc.height as f32);
+
+        let normal_view: &[f32; 16] = normal_view.as_ref();
+        let projection: &[f32; 16] = projection.as_ref();
+        let flipped_view: &[f32; 16] = flipped_view.as_ref();
+
+        let mut terrain_normal_uniform_data = Vec::<f32>::new();
+        let mut terrain_flipped_uniform_data = Vec::<f32>::new();
+
+        terrain_normal_uniform_data.extend_from_slice(projection);
+        terrain_normal_uniform_data.extend_from_slice(normal_view);
+        terrain_normal_uniform_data.extend_from_slice(&[0.; 4]);
+        terrain_flipped_uniform_data.extend_from_slice(projection);
+        terrain_flipped_uniform_data.extend_from_slice(flipped_view);
+        terrain_flipped_uniform_data.extend_from_slice(&[0., 1., 0., 0.0]);
+
+        let mut water_uniform_data = Vec::<f32>::new();
+
+        water_uniform_data.extend_from_slice(normal_view);
+        water_uniform_data.extend_from_slice(projection);
+        water_uniform_data.extend_from_slice(&[0.0, 0.0, SIZE]);
+
+        queue.write_buffer(&self.terrain_normal_uniform_buf, 0, bytemuck::cast_slice(&terrain_normal_uniform_data));
+        queue.write_buffer(&self.terrain_flipped_uniform_buf, 0, bytemuck::cast_slice(&terrain_flipped_uniform_data));
+        queue.write_buffer(&self.water_uniform_buf, 0, bytemuck::cast_slice(&water_uniform_data));
+    }
+
+    fn render(
+        &mut self,
+        frame: &wgpu::SwapChainTexture,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    ) -> wgpu::CommandBuffer {
+        // Increment frame count regardless of if we draw.
+        self.current_frame += 1;
+
+        // Write the sin/cos values to the uniform buffer for the water.
+        let (water_sin, water_cos) = ((self.current_frame as f32) / 400.0).sin_cos();
+        queue.write_buffer(&self.water_uniform_buf, std::mem::size_of::<[f32; 16]>() as wgpu::BufferAddress * 2, bytemuck::cast_slice(&[water_sin, water_cos]));
+
+        // The encoder provides a way to turn our instructions here, into
+        // a command buffer the GPU can understand.
+        let mut encoder =
+            device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+
+        // Only render valid frames. See resize method.
+        if self.active < self.current_frame {
```

Comment:
it doesn't explain what the crash is
also, consider doing an early out here: all the encoding code becomes indented otherwise